src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1357,185 +1362,220 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
       localA += null
-      val fn = rhs.eval(newContext)
-      val localFunctions = newContext.aggregationFunctions
+
+      val lambdaIdx = localA.length
+      localA += null
+      val bodyFn = body.eval(agg.ec.copy(st = agg.ec.st + (param -> (lambdaIdx, agg.elementType))))
+
+      val aggF = agg.f
       val (zv, seqOp, combOp) = {
-        val so = (sum: Any) => {
-          val counts = sum.asInstanceOf[(Long, Long)]
-          val ret = fn().asInstanceOf[Boolean]
-          if (ret)
-            (counts._1 + 1, counts._2 + 1)
-          else
-            (counts._1, counts._2 + 1)
+        val seqOp = (sum: Any, next: Any) => {
+          val (numerator, denominator) = sum.asInstanceOf[(Long, Long)]
+          val (value, include) = aggF(next)
+          if (include) {
+            localA(lambdaIdx) = value
+            val numToAdd = if (bodyFn().asInstanceOf[Boolean]) 1 else 0
+            (numerator + numToAdd, denominator + 1)
+          } else (numerator, denominator)
         }
         val co: (Any, Any) => Any = (left: Any, right: Any) => {
           val lh = left.asInstanceOf[(Long, Long)]
           val rh = right.asInstanceOf[(Long, Long)]
           (lh._1 + rh._1, lh._2 + rh._2)
         }
-        (() => (0L, 0L), so, co)
+        (() => (0L, 0L), seqOp, co)
       }
-      localFunctions += ((zv, seqOp, combOp, localIdx))
-      AST.evalCompose[Any](ec, lhs) { a =>
+      agg.ec.aggregationFunctions += ((zv, seqOp, combOp, localIdx))
+      () => {
         val (num: Long, denom: Long) = localA(localIdx)
         divNull(num.toDouble, denom)
       }
 
-
-    case (agg, "stats", Array(rhs)) =>
-      val newContext = agg.asInstanceOf[TAggregable].ec
-      val localA = newContext.a
+    case (agg: TAggregable, "stats", Array()) =>
+      val localA = agg.ec.a
       val localIdx = localA.length
       localA += null
-      val fn = rhs.eval(newContext)
-      val localFunctions = newContext.aggregationFunctions
 
-      val localType = rhs.`type`.asInstanceOf[TNumeric]
-      val seqOp = (a: Any) => {
-        val query = fn()
-        val sc = a.asInstanceOf[StatCounter]
-        if (query != null)
-          localType match {
-            case TInt => sc.merge(query.asInstanceOf[Int])
-            case TLong => sc.merge(query.asInstanceOf[Long])
-            case TFloat => sc.merge(query.asInstanceOf[Float])
-            case TDouble => sc.merge(query.asInstanceOf[Double])
-          }
-        else
+      val t = agg.elementType
+      val aggF = agg.f
+      val seqOp = (a: Any, next: Any) => {
+        val (value, included) = aggF(next)
+        if (included) {
+          val sc = a.asInstanceOf[StatCounter]
+          if (value != null)
+            t match {
+              case TInt => sc.merge(value.asInstanceOf[Int])
+              case TLong => sc.merge(value.asInstanceOf[Long])
+              case TFloat => sc.merge(value.asInstanceOf[Float])
+              case TDouble => sc.merge(value.asInstanceOf[Double])
```

Comment:
`StatCounter.merge` just takes a Double, so you might as well just call the numeric conversion code to Double here.